src/packages/packages.controller.ts

```diff
@@ -0,0 +1,198 @@
+import { 
+  Controller, 
+  Get, 
+  Param, 
+  StreamableFile, 
+  NotFoundException, Res
+} from '@nestjs/common';
+import { EntityManager } from '@mikro-orm/postgresql';
+import { Package } from '../model/package.entity';
+import { Response } from 'express';
+import * as stream from 'stream';
+import { Download } from 'src/model/download.entity';
+import { Version } from 'src/model/version.entity';
+
+@Controller('api/v1/packages')
+export class PackagesController {
+  constructor(private readonly em: EntityManager) {}
+
+  @Get(':name')
+  async getPackageName(@Param('name') name: string, @Param('version') version: string) {
+
+    const pkg = await this.em.findOne(Package, { name }, { populate: ['versions'] });
+    if (!pkg) {
+      throw new NotFoundException(`Package "${name}" not found`);
+    }
+
+    const ver = pkg.versions.getItems();
+
+    if (!ver) {
+      throw new NotFoundException(`Version "${version}" not found for package "${name}"`);
+    }
+
+    return {
+      name: pkg.name,
+      versions: ver,
+      readme: pkg.readme,
+      tag_name: pkg.tagName,
+    };
+  }
+
+  @Get(':name/:version')
+  async getPackage(@Param('name') name: string, @Param('version') version: string) {
+    const [major, minor, patch] = version.split('.').map(Number);
+
+    const pkg = await this.em.findOne(Package, { name }, { populate: ['versions'] });
```

Comment:
we should be able to fetch package by name and version here already in 1 DB query

Reply:
We have package name and version. We can do a fetch for specific package and version, not fetch whole versions. Pls fix it to avoid getting all versions. If you change primary keys to composite as I suggested above it will be index.